testInstallLargeSnapshotWithThrottle is flaky

### Describe the bug
Test `com.alipay.sofa.jraft.core.NodeTest#testInstallLargeSnapshotWithThrottle` is flaky. It is needed to run this test approxiamtely 20 times to get this error 

```
2021-08-11 19:16:57 [JRaft-Closure-Executor-7] ERROR Utils:180 - Fail to run done closure
java.lang.AssertionError: Status[ECATCHUP<10003>: Peer 172.25.4.112:5006 failed to catch up.]
	at org.junit.Assert.fail(Assert.java:89) ~[junit-4.13.1.jar:4.13.1]
	at org.junit.Assert.assertTrue(Assert.java:42) ~[junit-4.13.1.jar:4.13.1]
	at com.alipay.sofa.jraft.core.NodeTest.lambda$testInstallLargeSnapshotWithThrottle$1(NodeTest.java:2194) ~[test-classes/:?]
	at com.alipay.sofa.jraft.util.Utils$1.run(Utils.java:178) [classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_231]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_231]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_231]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_231]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_231]
```

The first time we encountered this failed test in our fork, we changed RPC layer in our repo, so as far as we reproduced this also in the original JRaft repo, I would say that this bug is not related to the network layer. I've also made some investigations on our side, I've recorded messages on the leader and we can see that right before the fail leader for some reason sent last two `AppendEntriesRequest` with an interval that is more than election timeout. As far as I can understand, that affects `lastRpcSendTimestamp` and further check for a catch-up does not pass and ends with an error 
```
Utils.monotonicMs() - this.replicatorGroup.getLastRpcSendTimestamp(peer) <= this.options
                    .getElectionTimeoutMs()
```

I can provide full debug logs, If it is needed

### Expected behavior
Node is successfully joined to the cluster with throttled snapshot and the test is not flaky

### Actual behavior
Test `testInstallLargeSnapshotWithThrottle` is flaky, and there is a chance that node cannot join the topology successfully 

### Steps to reproduce
Just run `testInstallLargeSnapshotWithThrottle` several time (~20) on the last matser revision (34af4e21cf976f7d5499634bf74f5686c5a70677).

### Environment

- SOFAJRaft version: 1.3.8
- JVM version (e.g. `java -version`):
java version "1.8.0_231"
Java(TM) SE Runtime Environment (build 1.8.0_231-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.231-b11, mixed mode)
- OS version (e.g. `uname -a`):
macOS Catalina 10.15.7
- Maven version:
Apache Maven 3.6.3
- IDE version:
Intellij IDEA 2019.3.5

--- a/jraft-core/src/main/java/com/alipay/sofa/jraft/util/ThreadId.java
+++ b/jraft-core/src/main/java/com/alipay/sofa/jraft/util/ThreadId.java
@@ -39,7 +39,7 @@ public class ThreadId {
 
     private final Object           data;
     private final NonReentrantLock lock                = new NonReentrantLock();
-    private final List<Integer>    pendingErrors       = Collections.synchronizedList(new ArrayList<>());
+    private final List<Integer>    pendingErrors       = new ArrayList<>();
     private final OnError          onError;
     private volatile boolean       destroyed;
 

--- a/jraft-core/src/main/java/com/alipay/sofa/jraft/util/ThreadId.java
+++ b/jraft-core/src/main/java/com/alipay/sofa/jraft/util/ThreadId.java
@@ -157,17 +157,19 @@ public class ThreadId {
         if (this.destroyed) {
             return;
         }
-        if (this.lock.tryLock()) {
-            if (this.destroyed) {
-                this.lock.unlock();
-                return;
+        synchronized (pendingErrors) {
+            if (this.lock.tryLock()) {
+                if (this.destroyed) {
+                    this.lock.unlock();
+                    return;
+                }
+                if (this.onError != null) {
+                    // The lock will be unlocked in onError.
+                    this.onError.onError(this, this.data, errorCode);
+                }
+            } else {
+                this.pendingErrors.add(errorCode);
             }
-            if (this.onError != null) {
-                // The lock will be unlocked in onError.
-                this.onError.onError(this, this.data, errorCode);
-            }
-        } else {
-            this.pendingErrors.add(errorCode);
         }
     }
 }

